### ISSUE TITLE
CollectionField iterator and ReferenceCollectionInput

### ISSUE THREAD

**Issue by joe-at-startupmedia** [marmelab/admin-on-rest#1662]
**What you were expecting:**

I want to populate a list iterator from a source of a collection. Not an array. And furthermore, I don't 
want to use a reference field. All of the data I need is contained in the collection. 

```
            <ReferenceArrayField label="Groups" source="roles">
                <SingleFieldList>
                    <ChipField source="name" />
                </SingleFieldList>
            </ReferenceArrayField>
```

Network Response:

```
[{
	"email": "bo.hatchet@mail.com",
	"id": 6,
	"roles": [{
		"id": 3,
		"name": "reg role"
	}]
}, {
	"email": "bo.ratchet@mail.com",
	"id": 7,
	"roles": [{
		"id": 1,
		"name": "admin"
	},{
		"id": 2,
		"name": "sales"
	}]
}]
```

**What happened instead:**

Currently there seems to be no iterator field to work on a source that is a collection. Only ReferenceArrayField and ReferenceManyField which both require and utilize a reference endpoint.

**Comment by joe-at-startupmedia** [comment]
It seems like the following documentation could be a solution to write a custom iterator for what I am trying to achieve.
https://github.com/marmelab/admin-on-rest/blob/master/docs/List.md#using-a-custom-iterator

I am wondering however, if this is a common occurence, would it be something you're willing to support as a builtin component. It seems like `<CollectionField>` would be the best name for it.

I would rather retrieve a list of 20 records with hydrated collection then retrieve a list of 20 records in one network request and then make 20 more network requests to populate the reference data. 1 request versus 21 network requests to retrieve the same information from the rest endpoint. Where this component would be most useful is in pagination, greatly reducing the amount of reference calls being made.

**Comment by joe-at-startupmedia** [comment]
Related to:
https://stackoverflow.com/questions/46592926/how-to-leverage-embedded-resources-in-referencearrayinput-and-referencearrayfiel

**Comment by joe-at-startupmedia** [comment]
After further investigation and learning this framework (which is super awesome by the way)  I realized that I am addressing two separate issues here. 

First issue is a field iterator for the lists. This turns out to be extremely simple and completely application specific so this is a mute issue. My example regarding this implementation is provided below.

```javascript
const CollectionField = (props) => (
    <div style={{ margin: '1em' }}>
    {props.record.roles.map(role =>
        <ChipField record={role} key={role.id} source="name" />
    )}
    </div>
);
```

The next issue which is addressed in the [stackoverflow post](https://stackoverflow.com/questions/46592926/how-to-leverage-embedded-resources-in-referencearrayinput-and-referencearrayfiel) I provided above related to the edit/create form inputs. Also, not so much of an issue because there are only a few foreign keys to request at load time.  This implementation is what I would relate to `ReferenceArrayInput` but once again this is somewhat more or less application specific. The only snafu I ran into this implementation is setting the default value from the source attribute hence all of the `defaultInputValue` spaghetti. My implementation is provided below.

```javascript
import { connect } from 'react-redux';
import { ReferenceArrayInput, translate } from 'admin-on-rest';
import {
    crudGetMany as crudGetManyAction,
    crudGetMatching as crudGetMatchingAction,
} from 'admin-on-rest/lib/actions/dataActions';
import { getPossibleReferences } from 'admin-on-rest/lib/reducer/admin/references/possibleValues';

const referenceSource = (resource, source) => `${resource}@${source}`;

let defaultInputValue = [];

function getDefaultInputValue(props) {
    return props.record[props.reference].map((r) => r.id);
}

export class ReferenceCollectionInput extends ReferenceArrayInput {
    constructor(props, context) {
        props.input.value = defaultInputValue = getDefaultInputValue(props);
        super(props, context);
    }
}

function mapStateToProps(state, props) {
    if(!props.input.value && defaultInputValue) {
      props.input.value = defaultInputValue;
      defaultInputValue = false;
    }
    const referenceIds = props.input.value || [];
    const data = state.admin.resources[props.reference].data;
    const matchingReferences = getPossibleReferences(
        state,
        referenceSource(props.resource, props.source),
        props.reference,
        referenceIds
    ).map((r) => r);
    const referenceRecords = referenceIds.reduce((references, referenceId) => {
        if (data[referenceId]) {
            references.push(data[referenceId]);
        }
        return references;
    }, []);
    return {matchingReferences, referenceRecords};
}

const ExportReferenceCollectionInput = connect(mapStateToProps, {
    crudGetMany: crudGetManyAction,
    crudGetMatching: crudGetMatchingAction,
})(translate(ReferenceCollectionInput));

ExportReferenceCollectionInput.defaultProps = {
    addField: true,
};

export default ExportReferenceCollectionInput;
```

```javascript
export const ReferrerEdit = (props) => (
    <Edit title="Edit Member" {...props}>
        <SimpleForm>
            <ReferenceCollectionInput source="role_ids" reference="roles" allowEmpty>
                <SelectArrayInput optionText="name" />
            </ReferenceCollectionInput>
        </SimpleForm>
    </Edit>
);
```

Two things wrong with this example. 
 - It seems that passing context in the constructor is functional but not encouraged. Have not seem many examples using this method. But it seems to be the only way to extend a component without duplicate code.
- As I've mentioned before the default value selection for the input derived from the source. It was not populated at render time. If anyone could provide any pointers to that, it would be greatly appreciated.

**Comment by fzaninotto** [comment]
Fixed by #1717
